### PR TITLE
Auto-generate mixin configs

### LINKED ISSUE #297: Don't initialize client mixins in a server environment
It seems that ModernFix tries to initialize all mixins regardless of the environment. On servers this causes the following annoying albeit harmless errors from attempting to load the client mixins:
```
[22:14:12] [main/WARN]: Error loading class: net/minecraft/class_340 (java.lang.ClassNotFoundException: net/minecraft/class_340)
[22:14:12] [main/WARN]: @Mixin target net.minecraft.class_340 was not found modernfix-fabric.mixins.json:feature.branding.GuiMixin from mod modernfix
[22:14:12] [main/WARN]: Error loading class: net/minecraft/class_1043 (java.lang.ClassNotFoundException: net/minecraft/class_1043)
[22:14:12] [main/WARN]: @Mixin target net.minecraft.class_1043 was not found modernfix-fabric.mixins.json:safety.DynamicTextureMixin from mod modernfix
[22:14:12] [main/WARN]: Error loading class: net/minecraft/class_310 (java.lang.ClassNotFoundException: net/minecraft/class_310)
[22:14:12] [main/WARN]: @Mixin target net.minecraft.class_310 was not found modernfix-fabric.mixins.json:core.MCMixin_Fabric from mod modernfix
[22:14:12] [main/WARN]: Error loading class: net/minecraft/class_634 (java.lang.ClassNotFoundException: net/minecraft/class_634)
[22:14:12] [main/WARN]: @Mixin target net.minecraft.class_634 was not found modernfix-fabric.mixins.json:core.ClientPlayNetHandlerMixin from mod modernfix
[22:14:12] [main/WARN]: Error loading class: net/minecraft/class_1071 (java.lang.ClassNotFoundException: net/minecraft/class_1071)
[22:14:12] [main/WARN]: @Mixin target net.minecraft.class_1071 was not found modernfix-common.mixins.json:perf.cache_profile_texture_url.SkinManagerMixin from mod modernfix
[22:14:12] [main/WARN]: Error loading class: net/minecraft/class_310 (java.lang.ClassNotFoundException: net/minecraft/class_310)
[22:14:12] [main/WARN]: @Mixin target net.minecraft.class_310 was not found modernfix-common.mixins.json:bugfix.world_leaks.MinecraftMixin from mod modernfix
[22:14:13] [main/WARN]: Error loading class: net/minecraft/class_528$class_4272 (java.lang.ClassNotFoundException: net/minecraft/class_528$class_4272)
[22:14:13] [main/WARN]: @Mixin target net.minecraft.class_528$class_4272 was not found modernfix-common.mixins.json:bugfix.world_screen_skipped.WorldSelectionListMixin from mod modernfix
[22:14:13] [main/WARN]: Error loading class: net/minecraft/class_287 (java.lang.ClassNotFoundException: net/minecraft/class_287)
[22:14:13] [main/WARN]: @Mixin target net.minecraft.class_287 was not found modernfix-common.mixins.json:bugfix.buffer_builder_leak.BufferBuilderMixin from mod modernfix
```

The mixins should be registered as per https://fabricmc.net/wiki/tutorial:mixin_registration#register_mixins, i.e. client mixins should be defined as `client` and server mixins as `server`, instead of a unifying them all.

### PR DESCRIPTION
**Pull request by Phoenix-Starlight**
Resolves #297 
1 (or 2?) ugly hack since AP output API doesn't have anything for resources
On the positive side, the directory scanning hack is dead.
- [x] Auto-generate mixin configs
- [x] Integrate AP
- [x] ~~Get victims to test whether or not it explodes~~
- [x] New: Emit compile-time errors for mixins that target client side classes but are not annotated with `@ClientOnlyMixin`
- [x] Fix remaining compile-time errors

### CODE REVIEW

**Comment by embeddedt** [comment]
I can test it easily by checking that it generates the same JSON as the previous approach. No victims required.

**Comment by Phoenix-Starlight** [comment]
RemapJar task originally didn't take the JSON. (Hence the hack into resource sourceset)

**Comment by Phoenix-Starlight** [comment]
Ready. (Merge conflicts incoming for later branches...)

File: forge/src/main/java/org/embeddedt/modernfix/forge/mixin/perf/async_locator/EyeOfEnderAccess.java [hunk 1]
```diff
@@ -1,11 +1,13 @@
 package org.embeddedt.modernfix.forge.mixin.perf.async_locator;
 
 import net.minecraft.world.entity.projectile.EyeOfEnder;
+import org.embeddedt.modernfix.annotation.ClientOnlyMixin;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.gen.Accessor;
 
 @Mixin(EyeOfEnder.class)
+@ClientOnlyMixin
```
Comment (embeddedt) [inline forge/src/main/java/org/embeddedt/modernfix/forge/mixin/perf/async_locator/EyeOfEnderAccess.java]: The async locator mixins should not be client-only.

**Review by embeddedt — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/groovy/modernfix.mod-common-conventions.gradle [hunk 2]
```diff
@@ -16,37 +16,26 @@ dependencies {
...
+    }
+}
 
-        inputs.properties replacements
-        def filePattern = "modernfix-" + project.name + ".mixins.json"
-        filesMatching(filePattern) {
-            expand replacements
+tasks.withType(JavaCompile) {
+    options.fork = true
+    options.forkOptions.jvmArgs << '--enable-preview'
```
Comment (embeddedt) [inline buildSrc/src/main/groovy/modernfix.mod-common-conventions.gradle:35]: Why does the mod itself need to be compiled with `--enable-preview`? It conflicts with the commit description, which says it's only needed for the AP: https://github.com/embeddedt/ModernFix/commit/4645c2622742d62ecf9fd08028897edb40512be3

**Review by embeddedt — CHANGES_REQUESTED** [review]
Very good work; I'm very happy with this. Tested locally and the generated mixin list is identical to my original hack, except it properly splits client mixins into the client list.

I just have one last question before merging.

(At some point, we might consider moving this into a standalone plugin so that it can be used by other projects.)

File: buildSrc/src/main/groovy/modernfix.mod-common-conventions.gradle (hunk 2 above)
Comment (Phoenix-Starlight) [reply buildSrc/src/main/groovy/modernfix.mod-common-conventions.gradle:35]: So the AP itself is compiled with preview... but you need to enable preview features on the JVM to run the AP. This hunk specifically enables it for the compilation process/AP but not the code being compiled.
See https://github.com/gradle/gradle/issues/12680

**Pull request merged**